Mathlib/Algebra/Category/ModuleCat/Presheaf/Sheafification.lean

```diff
@@ -0,0 +1,119 @@
+  obj M₀ := sheafify α (CategoryTheory.toSheafify J M₀.presheaf)
+  map f := sheafifyMap _ _ _ f ((presheafToSheaf J AddCommGroupCat).map f.hom) (by simp)
+  map_id M₀ := by
+    ext1
+    apply (toPresheaf _).map_injective
+    simp [toPresheaf, sheafify]
+  map_comp _ _ := by
+    ext1
+    apply (toPresheaf _).map_injective
+    simp [toPresheaf, sheafify]
+
+attribute [-simp] sheafification_map
+
+/-- The sheafification of presheaves of modules commutes with the functor which
+forgets the module structures. -/
+noncomputable def sheafificationCompToSheaf :
+    sheafification.{v} α ⋙ SheafOfModules.toSheaf _ ≅
+      toPresheaf _ ⋙ presheafToSheaf J AddCommGroupCat :=
+  Iso.refl _
+
+/-- The sheafification of presheaves of modules commutes with the functor which
+forgets the module structures. -/
+noncomputable def sheafificationCompForgetCompToPresheaf :
+    sheafification.{v} α ⋙ SheafOfModules.forget _ ⋙ toPresheaf _ ≅
+      toPresheaf _ ⋙ presheafToSheaf J AddCommGroupCat ⋙ sheafToPresheaf J AddCommGroupCat :=
+  Iso.refl _
+
+/-- The bijection between types of morphisms which is part of the adjunction
+`sheafificationAdjunction`. -/
+noncomputable def sheafificationHomEquiv
+    {P : PresheafOfModules.{v} R₀} {F : SheafOfModules.{v} R} :
+    ((sheafification α).obj P ⟶ F) ≃
+      (P ⟶ (restrictScalars α).obj ((SheafOfModules.forget _).obj F)) := by
+  apply sheafifyHomEquiv
+
+lemma sheafificationHomEquiv_hom'
+    {P : PresheafOfModules.{v} R₀} {F : SheafOfModules.{v} R}
+    (f : (sheafification α).obj P ⟶ F) :
+    (sheafificationHomEquiv α f).hom =
+      CategoryTheory.toSheafify J P.presheaf ≫ f.val.hom := rfl
+
+lemma sheafificationHomEquiv_hom
+    {P : PresheafOfModules.{v} R₀} {F : SheafOfModules.{v} R}
+    (f : (sheafification α).obj P ⟶ F) :
+    (sheafificationHomEquiv α f).hom =
+      (sheafificationAdjunction J AddCommGroupCat).homEquiv P.presheaf
+        ((SheafOfModules.toSheaf _).obj F) ((SheafOfModules.toSheaf _).map f) := by
+  rw [sheafificationHomEquiv_hom', Adjunction.homEquiv_unit]
+  dsimp
+
+lemma toSheaf_map_sheafificationHomEquiv_symm
+    {P : PresheafOfModules.{v} R₀} {F : SheafOfModules.{v} R}
+    (g : P ⟶ (restrictScalars α).obj ((SheafOfModules.forget _).obj F)) :
+    (SheafOfModules.toSheaf _).map ((sheafificationHomEquiv α).symm g) =
+      (((sheafificationAdjunction J AddCommGroupCat).homEquiv
+        P.presheaf ((SheafOfModules.toSheaf R).obj F)).symm g.hom) := by
+  obtain ⟨f, rfl⟩ := (sheafificationHomEquiv α).surjective g
+  apply ((sheafificationAdjunction J AddCommGroupCat).homEquiv _ _).injective
+  rw [Equiv.apply_symm_apply, Adjunction.homEquiv_unit, Equiv.symm_apply_apply]
+  rfl
+
+/-- Given a locally bijective morphism `α : R₀ ⟶ R.val` where `R₀` is a presheaf of rings
+and `R` a sheaf of rings, this is the adjunction
+`sheafification.{v} α ⊣ SheafOfModules.forget R ⋙ restrictScalars α`. -/
+noncomputable def sheafificationAdjunction :
```

Comment:
Can you provide simp lemmas for this?

Reply:
I will add a lemma for the `homEquiv` field of the adjunction. However, I think we need to fix an issue with the general `Adjunction` API: the lemmas `Adjunction.homEquiv_(co)unit` expressing `homEquiv` and `homEquiv.symm` in terms of the `unit`  and `counit` are simp lemmas, which is problematic for adjunctions that are mainly defined using `Adjunction.mkOfHomEquiv`.

Reply:
Doesn't the units and counits have good descriptions? (`toSheafify` and some isomorphism)

Reply:
I have added lemmas for `homEquiv` and `unit`.

Reply:
Will we get the description of the counit later in the future?

Reply:
I do not think so, as I am not sure there is much to say about the counit.

Reply:
I was thinking about something along the lines of 
```lean
def forgetRestrictScalarsFullyFaithful [Presheaf.IsLocallySurjective J α] :
    (SheafOfModules.forget.{v} R ⋙ restrictScalars α).FullyFaithful := by
  fconstructor
  · intro M N f
    refine ⟨f.1, ?_⟩
    intros U r x
    have := Presheaf.isSheaf_comp_of_isSheaf J N.1.presheaf (forget _) N.2
    have := ((isSheaf_iff_isSheaf_of_type _ _).mp this _ (Presheaf.imageSieve_mem J α r))
    apply this.isSeparatedFor.ext
    rintro V i ⟨r' : R₀.obj (.op V), hr'⟩
    simp only [Functor.comp_obj, Opposite.op_unop, SheafOfModules.forget_obj, restrictScalars_obj,
      Functor.comp_map, AddCommGroupCat.forget_map, PresheafOfModules.map_smul, ← hr']
    erw [← NatTrans.naturality_apply f.hom i.op, ← NatTrans.naturality_apply f.hom i.op,
      ← f.map_smul, PresheafOfModules.map_smul] -- bad
    rw [← hr']
    rfl
  · exact fun f ↦ rfl
  · exact fun f ↦ rfl

example [Presheaf.IsLocallySurjective J α] : IsIso (sheafificationAdjunction α).counit :=
  letI := (forgetRestrictScalarsFullyFaithful α).faithful
  letI := (forgetRestrictScalarsFullyFaithful α).full
  inferInstance
```

Reply:
Yes, apart from saying that the counit is an iso, we cannot say much :-)